api/client/cp.go

```diff
@@ -1,57 +1,253 @@
+// so we have to check for a `/` or `.` prefix. Also, in the case of a Windows
+// client, a `:` could be part of an absolute Windows path, in which case it
+// is immediately proceeded by a backslash.
+func splitCpArg(arg string) (container, path string) {
+	if filepath.IsAbs(arg) {
+		// Explicit local absolute path, e.g., `C:\foo` or `/foo`.
+		return "", arg
 	}
-	if statusCode == 404 {
-		return fmt.Errorf("No such container: %v", info[0])
+
+	parts := strings.SplitN(arg, ":", 2)
+
+	switch {
+	case len(parts) == 1:
+		// No `:` in the arg.
+		fallthrough
+	case strings.HasPrefix(parts[0], "."):
+		// Explicit local relative path.
+		return "", arg
+	default:
+		return parts[0], parts[1]
+	}
+}
+
+type containerPathStat struct {
+	Name  string      `json:"name"`
+	Path  string      `json:"path"`
+	Size  int64       `json:"size"`
+	Mode  os.FileMode `json:"mode"`
+	Mtime time.Time   `json:"mtime"`
+}
+
+func (cli *DockerCli) statContainerPath(containerName, path string) (containerPathStat, error) {
+	var stat containerPathStat
+
+	query := make(url.Values, 1)
+	query.Set("path", filepath.ToSlash(path)) // Normalize the paths used in the API.
+
+	urlStr := fmt.Sprintf("/containers/%s/stat-path?%s", containerName, query.Encode())
+
+	response, _, statusCode, err := cli.call("GET", urlStr, nil, nil)
+	if err != nil {
+		return stat, err
+	}
+
+	defer response.Close()
+
+	if statusCode != 200 {
+		return stat, fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	err = json.NewDecoder(response).Decode(&stat)
+
+	return stat, err
+}
+
+func resolveLocalPath(localPath string) (absPath string, err error) {
+	if absPath, err = filepath.Abs(localPath); err != nil {
+		return
+	}
+
+	return archive.PreserveTrailingDotOrSeparator(absPath, localPath), nil
+}
+
+func (cli *DockerCli) copyFromContainer(srcContainer, srcPath, dstPath string) (err error) {
+	if dstPath != "-" {
+		// Get an absolute destination path.
+		dstPath, err = resolveLocalPath(dstPath)
+		if err != nil {
+			return err
+		}
 	}
+
+	query := make(url.Values, 1)
+	query.Set("path", filepath.ToSlash(srcPath)) // Normalize the paths used in the API.
+
+	urlStr := fmt.Sprintf("/containers/%s/archive-path?%s", srcContainer, query.Encode())
+
+	response, hdr, statusCode, err := cli.call("GET", urlStr, nil, nil)
 	if err != nil {
 		return err
 	}
 
-	hostPath := cmd.Arg(1)
-	if statusCode == 200 {
-		if hostPath == "-" {
-			_, err = io.Copy(cli.out, stream)
-		} else {
-			err = archive.Untar(stream, hostPath, &archive.TarOptions{NoLchown: true})
+	defer response.Close()
+
+	if statusCode != 200 {
+		return fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	if dstPath == "-" {
+		// Send the response to STDOUT.
+		_, err = io.Copy(os.Stdout, response)
+
+		return err
+	}
+
+	// Prepare source copy info.
+	srcInfo := archive.CopyInfo{
+		Path:   srcPath,
+		Exists: true,
+		IsDir:  hdr.Get("X-Docker-Container-Path-Is-Dir") == "true",
```

Comment:
Nice to see lot's of discussion on this. :smiley: 

The `extract-to-dir` endpoint requires the destination path to be a directory because a tar archive is always extracted in the context of a directory. The "copy a single file" use case works correctly because `docker cp` is smart enough to "rebase" the source archive to match the filename of the destination and extract it to the parent directory of the target path instead, resulting in the expected behavior. It's admittedly a hack, but it's a well-documented, cp-only hack with lots of comments in `pkg/archive/copy.go`.

@tonistiigi said:

> I want all these hacks to be on the client side and none of them in the API side. Most of them were already moved, a bit still left for my taste.

Which are the remaining server-side hacks you are referring to? I'm willing to remove it. If you guys think the `X-Docker-Container-Path-Is-Dir` header is a hack (which it arguably is) I can remove it, but this would require an extra `stat-path` call before making the `archive-path` call, or I could also do what we discussed earlier: encode all of the stat-path info into a response header for `archive-path`. Then calling `HEAD archive-path` could be functionally equivalent to `GET stat-path` and we can remove the `stat-path` endpoint.

**However**, once we start talking about glob patterns, `stat` headers no longer make any sense because there could be multiple matches! Which does the `stat` apply to? If the "copy-from-a-container" logic in `docker cp` decided to call `GET stat-path` first with a glob pattern what would happen? Perhaps it would make sense to also have a `X-Docker-Container-Path-Multi=true` header to indicate that there were multiple matches?

For this reason I think it's bet to keep the `X-Docker-Container-Path-Is-Dir` header.